### ISSUE TITLE
Add exclude in the query arguments

### ISSUE THREAD

**Issue by pantharshit00** [prisma/prisma-client-js#649]
## Problem
Sometimes it is helpful to just remove some fields from the selection set like `password` so that we accidentally don't expose them to the frontend.

## Solution
Adding a `exclude` key in the arguments of the queries to easily exclude fields from the selectionSet just like `include` or `select`.

```js
prisma.user.findMany({
  exclude:{
    password: true
  }
})
```

## Alternatives
1. Schema annotation
2. Manual omission via lodash or something !?

## Related
https://github.com/prisma/prisma/issues/2170

**Comment by thecodeah** [comment]
@pantharshit00 Despite our suggestions being very similar, I still believe the method I had suggested is better. Instead of having to remember to exclude certain fields every single time we fetch something (Which can *still* accidentally be forgotten) you specify it once in the Prisma Schema and explicitly request it when it's needed.

https://github.com/prisma/prisma/issues/2498

**Comment by pantharshit00** [comment]
Hey @thecodeah, I think I linked the same thing twice in your issue. Added a note in your issue: https://github.com/prisma/prisma/issues/2498#issuecomment-630722344

**Comment by joaopaulobdac** [comment]
At least for me, this is a very wanted feature now because TS forbids using the delete operator since TS 4.0:

https://github.com/microsoft/vscode/issues/96022#issuecomment-635729089

![20200825184009](https://user-images.githubusercontent.com/62244135/91230857-ac023c80-e702-11ea-9c79-95533aa5c9e1.png)

As a workaround I'm using the object rest syntax `{ password, ...userWithoutPassword } = user`, but the unused variable warnings are kinda annoying.

**Comment by darioielardi** [comment]
This becomes an even bigger problem when the column I want to hide is in a nested object as a result of an `include` query.

I think an `omit` clause would work but a schema annotation would be even better.

**Comment by matthewmueller** [comment]
@thecodeah, @darioielardi, and @flybayer expressed concerns with this proposal. Each query you write needs to explicitly exclude private fields. They would prefer an **excluded-by-default** approach.

To complicate matters, being able to access excluded fields from the Prisma Client is important for tests.

One idea proposed by @darioielardi is to add an attribute in the schema that will exclude a field. @flybayer suggested that if you want to include an excluded-by-default field (for tests) you could select that field explicitly. 

Concretely, that would look like this:

```prisma
model User {
  id Int @id
  password String @hidden // or @private
}
```

```js
prisma.user.findFirst({})
// { id: 1 }

prisma.user.findFirst({
  select: { 
    password: true
  }
})
// { password: "8674305" }
```

The downside of this is that when you do want to include an excluded field, you need to explicitly select all the other fields, like `id`.

Another possibility is excluding fields when you initialize the client. Something like:

```ts
const prisma = new PrismaClient({
  exclude: { password: true }
})
```

---

What do you think? Do you have any other ideas on how we could solve this?

**Comment by flybayer** [comment]
I would prefer it in the schema so you can see everything in one view and easily see if you forgot to hide a certain field. Whereas if it's in another file (and one that is rarely viewed), you can easily forget about it.

**Comment by darioielardi** [comment]
I agree with @flybayer, it would definitely prefer to have it in the schema.

However I'm concerned about the need to add *every* field to the `select` object to explicitly select a hidden column. 

What about something like an additional `expose` field exclusively aimed to select hidden columns? Something like:

```prisma
model User {
  id Int @id
  name String
  password String @hidden
}
```
```typescript
prisma.user.findMany({
  expose: { password: true }, // no need to explicitly select the other fields
})
```

That's probably a bit too much, but it might work.

**Comment by Kamshak** [comment]
I think a feature where you list the fields that are included by default would be better than a list of fields that are excluded, because as you add new fields to the db someone will forget to add @hidden and accidentally expose it in some API that simply returns say the result of the findOne as JSON. Maybe you could instead explicitly list the fields that are returned by default? Adding to that list will make it intentional, whereas when just adding a field to the DB you probably didn't want to change your API.

You could also leave the decision (list include or list exclude) up to the developer. Algolia does this for example here: https://www.algolia.com/doc/api-reference/api-parameters/attributesToRetrieve/. On an "index" which is basically a table you can specify which fields will be retreived by the client by default. It can be overwritten at query level. Here is how they solved it:

```
'attributesToRetrieve' => [
  'attribute1', // list of attributes to retrieve
  'attribute2'
]
'attributesToRetrieve' => [
  '*' // retrieves all attributes
]
'attributesToRetrieve' => [
  '*', // retrieves all attributes
  '-attribute1', // except this list of attributes (starting with a '-')
  '-attribute2'
]
``` 

Maybe this approach could work as well?

**Comment by scriptcoded** [comment]
I think @Kamshak's point about forgetting to add `@hidden` to a field makes sense to some degree, although at the same time I believe that the majority of fields on any given model will not be hidden. This means that for every field added you also have to explicitly mark those fields as non-hidden.

> ```
> const prisma = new PrismaClient({
>   exclude: { password: true }
> })
> ```
I also believe that there is an issue with this method of excluding fields suggested by @matthewmueller. There might might be multiple models with the same field, but all models shouldn't necessarily have them be hidden. Also, it might become unclear that a field is hidden when creating a new model and the creator of that model is unaware of a field being excluded "globally". 

The one time I could see this being useful would be when you want to exclude something on all models, like `createdAt`.

Personally I think adding a `@hidden` attribute makes a lot more sense as it reduces the amount of code necessary for each model, keeps the option at model level, and is only adding an attribute which makes it backwards compatible.

**Comment by samrith-s** [comment]
While this is good, we would still have to manually exclude things everywhere we use it, right?

Or does this mean we simply override parameters in the middleware?

**Comment by YannSuissa** [comment]
Any progress on this exclude issue? 
I'm struggling to hide a private field.. on client side it's easy to hide it but it you use a tool like playground it's easy to hack.
An `exclude` keyword would be perfect.
Thanks

**Comment by benjibuiltit** [comment]
Came here looking for a similar feature. I'll add my input that I'd also like to see an attribute at the model level and then an explicit opt-in when necessary to avoid any accidental leaking.

I'll also offer up an alternative workaround for those who may stumble across this issue before the feature is released. You can create a model for the sensitive data which you want to hide and use a relation to map between the main model and the sensitive data. The sensitive data will live in an alternate table which you can query for when you do need it, but your query for your main model won't have to be "massaged" so frequently. For example:

```
model Users {
  id  String  @id @default(uuid())
  email String
}

model HashedPasswords {
  userId String @id
  user Users @relation(fields: [userId], references: [id])
  password String
}
```

This will allow you to execute a nested query w/ a transactional guarantee that your sensitive data is written/created w/ your main model.
```
const result = await prisma.hashedPasswords.create({
      data: {
        password: hashedPassword,
        user: {
          create: {
            email
          }
        }
      }
    });
```

Because the sensitive data exists in a separate table you can query your main table w/o worrying about leaking anything.

And finally, to explicitly fetch the password for validation...
```
const password = await prisma.hashedPasswords.findUnique({
      where: {
        userId: {
          equals: userId
        }
      }
})
```
